### PR TITLE
Document admin bootstrapping and recovery

### LINKED ISSUE #30011: Document admin bootstrapping and recovery
The documentation should cover:
* The [`kc.sh|bat bootstrap-admin user`](https://github.com/keycloak/keycloak/issues/29324) and [`kc.sh|bat bootstrap-admin service`](https://github.com/keycloak/keycloak/issues/30002) commands.
* The [startup env vars](https://github.com/keycloak/keycloak/issues/30003).
* [Operator](https://github.com/keycloak/keycloak/issues/30004).
* That admin recovery can be performed through admin bootstrapping.
  * Example of how realms with increased security (OTP requirement, passwordles, ...) can use the admin service account to recover access through Admin CLI. 
* Release notes and migration notes for startup env vars.

### PR DESCRIPTION
**Pull request by Pepo48**
Closes: #30011

Signed-off-by: Peter Zaoral <pzaoral@redhat.com>

### CODE REVIEW

File: docs/guides/server/bootstrap-admin-recovery.adoc [hunk 1]
```diff
@@ -0,0 +1,90 @@
+<#import "/templates/guide.adoc" as tmpl>
+<#import "/templates/kc.adoc" as kc>
+<#import "/templates/links.adoc" as links>
+
+<@tmpl.guide
+title="Admin bootstrap and recovery"
+summary="Learn how to bootstrap and recover admin account." >
```
Comment (Pepo48) [inline docs/guides/server/bootstrap-admin-recovery.adoc:7]: As I mentioned during the meeting, initially I tried: 
```suggestion
summary="Learn how to bootstrap and recover admin account." 
includedOptions="bootstrap-admin-*">
```
but the Relevant options table ended up to be empty. Any idea why?

File: docs/guides/server/bootstrap-admin-recovery.adoc [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+== A solution for a disaster scenario
+
+It used to be difficult to regain access to a {project_name} instance, where all admin users have been locked out. The process required multiple user-advanced steps, including direct database access and manual changes. In the continuous effort to improve the user experience, {project_name} now provides multiple ways to bootstrap a new admin account, which can be later used to recover from such a situation.
```
Comment (vmuzikar) [inline docs/guides/server/bootstrap-admin-recovery.adoc:11]: This paragraph is good but possibly should be moved to release notes. IMHO guides should not reminiscence on some previous state of things.

Instead, I'd suggest some introduction that would explain the user the concept of temporary admin user (that they should create a more permanent and secure user etc.).

File: docs/guides/server/bootstrap-admin-recovery.adoc [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+<@kc.bootstrapadmin parameters="user"/>
+
+This command can be executed even before the {project_name} first ever start. In other words, it does not matter, whether the `master` realm already exists or not. If no other parameters are specified and/or no corresponding environment variables are set, a user is prompted to enter required information. For more information, see the <<Environment variables>> section bellow.
+
+Alternatively, the parameters can be directly specified in the command:
+
+<@kc.bootstrapadmin parameters="user --username tmpadm --password tmpadmpass"/>
+
+This command creates a temporary admin user with the username `tmpadm` and the password `tmpadmpass`. The user is valid for 60 minutes.
```
Comment (vmuzikar) [inline docs/guides/server/bootstrap-admin-recovery.adoc:25]: We do not automatically expire the user at this phase, see https://github.com/keycloak/keycloak/issues/30008#issuecomment-2151956072.
```suggestion
This command creates a temporary admin user with the username `tmpadm` and the password `tmpadmpass`.
```

File: docs/guides/server/bootstrap-admin-recovery.adoc [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+Passwordless, OTP or other advanced authentication methods can be enforced for the realm with lost admin access. In such case, the admin service account needs to be created. After the service account is created, authentication against the {project_name} instance is required to perform all the necessary operations:
+
+[source]
+----
+POST realms/{realm}/protocol/openid-connect/token \
+  -H "Content-Type: application/x-www-form-urlencoded" \
+  -d "client_id=<service_account_id>" \
+  -d "client_secret=<service_account_secret>" \
+  -d "grant_type=client_credentials"
```
Comment (vmuzikar) [inline docs/guides/server/bootstrap-admin-recovery.adoc:53]: We could probably use kcadm.sh|bat in these examples right away.

File: docs/guides/server/bootstrap-admin-recovery.adoc [hunk 5]
```diff
@@ -0,0 +1,90 @@
+<#import "/templates/guide.adoc" as tmpl>
+<#import "/templates/kc.adoc" as kc>
+<#import "/templates/links.adoc" as links>
+
+<@tmpl.guide
+title="Admin bootstrap and recovery"
+summary="Learn how to bootstrap and recover admin account." >
+
+== A solution for a disaster scenario
```
Comment (vmuzikar) [inline docs/guides/server/bootstrap-admin-recovery.adoc:9]: Bootstrapping is not just for disaster recovery scenarios.

File: docs/guides/server/bootstrap-admin-recovery.adoc [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+Alternatively, you can specify the parameters directly in the command:
+
+<@kc.bootstrapadmin parameters="service --client-id tmpclient --client-secret:env=SECRET_VAR"/>
+
+This command creates a temporary admin service account with the client ID `tmpclient` and the secret retrieved from the environment variable.
+
+== Regaining access to the realm with an increased security
+
+Passwordless, OTP or other advanced authentication methods can be enforced for the realm with lost admin access. In such case, the admin service account needs to be created. After the service account is created, authentication against the {project_name} instance is required to perform all the necessary operations:
```
Comment (vmuzikar) [inline docs/guides/server/bootstrap-admin-recovery.adoc:45]:
```suggestion
Passwordless, OTP or other advanced authentication methods can be enforced for the realm with lost admin access. In such case, the admin service account needs to be created to recover lost admin access to such realm. After the service account is created, authentication against the {project_name} instance is required to perform all the necessary operations:
```

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (vmuzikar) [inline docs/guides/server/bootstrap-admin-recovery.adoc:1]: It'd be also good to add a section for bootstrapping at server startup. Explaining the difference from the dedicated command (that on startup master realm must on exist, making the dedicated command the only suitable option for recovery).

We should also add some note that the dedicated command requires the server to be stopped and that it requires the same args as start (db options etc.).

**Review by vmuzikar — COMMENTED** [review]
@Pepo48 Thank you for the draft. I think it's a good start.

File: docs/guides/server/bootstrap-admin-recovery.adoc (hunk 3 above)
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc:25]: @vmuzikar you probably reviewed an older revision - at the time you posted the review the text didn't contain any reference regarding the expiration. I immediately cleaned it up after our discussion at the meeting.

Thanks for pointing it out in the fist place.

Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: This was a leftover I forgot about. Adressed, thanks!

File: docs/guides/server/bootstrap-admin-recovery.adoc (hunk 4 above)
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: Adressed.

File: docs/guides/server/bootstrap-admin-recovery.adoc (hunk 5 above)
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: Adressed.

File: docs/guides/server/bootstrap-admin-recovery.adoc (hunk 6 above)
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: Adressed.

File: docs/guides/server/bootstrap-admin-recovery.adoc [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+If no other parameters are specified and/or no corresponding environment variables are set, a user is prompted to enter required information. For more information, see the <<Environment variables>> section bellow.
+
+Alternatively, the parameters can be directly specified in the command:
+
+<@kc.bootstrapadmin parameters="user --username tmpadm --password tmpadmpass"/>
+
+This command creates a temporary admin user with the username `tmpadm` and the password `tmpadmpass`.
+
+NOTE: The command can be executed even before the {project_name} first ever start. However, it leaves the {project_name} server in a state, where the temporary admin account will be the only account that exists in the master realm - the default admin will not be created. If the intention is to recover a lost admin access, the default admin should exist within the master realm prior to the command execution. Also, it is strongly recommended to use the dedicated command with the very sane options that the {project_name} server is started with (e.g. `db` options).
```
Comment (Pepo48) [inline docs/guides/server/bootstrap-admin-recovery.adoc:27]: > It'd be also good to add a section for bootstrapping at server startup. Explaining the difference from the dedicated command (that on startup master realm must on exist, making the dedicated command the only suitable option for recovery).

I tried to address this one via a note.
@shawkins wdyt?

File: docs/guides/server/bootstrap-admin-recovery.adoc [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+If no corresponding environment variable is set, the command will fail with an error message, because the required password parameter is missing.
+
+== Environment variables
+
+For the `bootstrap-admin user` command, both username and password can be optionally set as environment variables:
+
+<@kc.bootstrapadmin parameters="user --username:env=<YourUsernameEnv> --password:env=<YourPassEnv>"/>
+
+Regarding the `bootstrap-admin service` command, for the client ID it is optional, while for the client secret it is required to be set as an environment variable:
```
[the remark this answers is not in the document]
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: Addressed. Thanks @shawkins.

**Comment by Pepo48** [comment]
I tried to incorporate every suggestion mentioned above. For now the changes are in a separate commit in order to make it easier to see the differences. I will do the squash later.

cc: @shawkins

**Review by shawkins — DISMISSED** [review]
> I tried to incorporate every suggestion mentioned above. For now the changes are in a separate commit in order to make it easier to see the differences. I will do the squash later.
> 
> cc: @shawkins
> 
> Edit: I also tried to adjust the [punctuation, articles and wording](https://github.com/keycloak/keycloak/compare/4f47c6e66396f099aa240643de797a26ddb16b6d..ff2d3d9d81f8bb7790cb5bbe265209031b032df2).

Thanks @Pepo48 it looks good. Just a conflict to clear up and it should be set.

**Comment by Pepo48** [comment]
> Just a conflict to clear up and it should be set.

@shawkins thanks for the review. The conflict happened in the release notes - I merged both of our inputs together into one section.

**Review by shawkins — DISMISSED** [review]
> > Just a conflict to clear up and it should be set.
> 
> @shawkins thanks for the review. The conflict happened in the release notes - I merged both of our inputs together into one section.

Looks good. Do you want to wait for @vmuzikar to also review?

**Comment by Pepo48** [comment]
@shawkins last week I was told that the whole bootstrap admin and recovery initiative is not considered as a blocker for the upcoming major release. That being said, we don't have to rush it, personally I just wanted to have the bits and pieces ready when they return. (to be able to react quickly then, if for whatever reason necessary)

tl;dr I'm fine with both options 😊

File: docs/guides/server/bootstrap-admin-recovery.adoc [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+If no other parameters are specified and/or no corresponding environment variables are set, a user is prompted to enter the required information. For more information, see the <<Environment variables>> section below.
+
+Alternatively, the parameters can be directly specified in the command:
+
+<@kc.bootstrapadmin parameters="user --username tmpadm --password tmpadmpass"/>
+
+This command creates a temporary admin user with the username `tmpadm` and the password `tmpadmpass`.
+
+NOTE: The command can be executed even before the first-ever start of {project_name}. However, this will leave the {project_name} server in a state where the temporary admin account is the only account that exists in the master realm and the default admin will not be created. The initial admin creation happens only when the master realm is created. If the intention is to recover lost admin access, the default admin must exist in the master realm prior to executing the command (see <@links.server id="configuration" anchor="create-initial-admin"/>). Additionally, it is strongly recommended to use the dedicated command with the same options that the {project_name} server is started with (e.g., `db` options).
```
Comment (ahus1) [inline docs/guides/server/bootstrap-admin-recovery.adoc:27]: IMHO the `db` options would not only be "strongly recommended", but "required" as it would otherwise not write to the correct database and it will not work. 

At the same time I am not sure if all of the options I use on the start command will also work on the boostrap commands (like: setting a HTTP port).

**Comment by ahus1** [comment]
@Pepo48 - I pushed one commit to make the external link check pass. I also found that anchors in the IDs don't work, so I updated the Freemarker makros to support it. It is already part of the scaling guide, but not yet in main. 

If you consider this documentation ready-to-be-merged, please ask Andrew Munro for a review from the technical writer's perspective to make sure the wording aligns with all the other docs.

File: docs/guides/server/bootstrap-admin-recovery.adoc (hunk 9 above)
Comment (shawkins) [reply docs/guides/server/bootstrap-admin-recovery.adoc:27]: This is a somewhat confusing area in general for us #29200 - we do some configuration exclusions for the CLI, but those same options will still work via the other config sources (the operator effectively uses this for export because almost everything is set via an env variable). Only a handful of options have hard-coded settings when running a non-server command.

Comment (shawkins) [reply docs/guides/server/bootstrap-admin-recovery.adoc:27]: I tried to clarify the cases with https://github.com/keycloak/keycloak/issues/29200#issuecomment-2233350379 - as there are potentially 3 different ways users could be invoking non-server commands which have some nuance that we trying to address here quickly with this recommendation.

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (vmuzikar) [reply docs/guides/server/bootstrap-admin-recovery.adoc:1]: > We should also add some note that the dedicated command requires the server to be stopped and that it requires the same args as start (db options etc.).

I believe this is not resolved yet.

File: docs/guides/server/bootstrap-admin-recovery.adoc [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+To disable the prompt for the parameters, the `--no-prompt` parameter can be used. For example:
+
+<@kc.bootstrapadmin parameters="user --username tmpadm --no-prompt"/>
+
+If no corresponding environment variable is set, the command will fail with an error message indicating that the required password parameter is missing.
+
+== Environment variables
+
+For the `bootstrap-admin user` command, both username and password can be optionally set as environment variables:
```
Comment (vmuzikar) [inline docs/guides/server/bootstrap-admin-recovery.adoc:79]: > the client ID is optional and defaults to temp-admin-service

Username also has a default value. Please see the comment above.

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@Pepo48 Thank you for the updates. Nice work! Added some comments.

File: docs/guides/server/bootstrap-admin-recovery.adoc [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+To create a temporary admin user, execute the following command:
+
+<@kc.bootstrapadmin parameters="user"/>
+
+If no other parameters are specified and/or no corresponding environment variables are set, a user is prompted to enter the required information. For more information, see the <<Environment variables>> section below.
+
+Alternatively, the parameters can be directly specified in the command:
+
+<@kc.bootstrapadmin parameters="user --username tmpadm --password tmpadmpass"/>
```
[the remark this answers is not in the document]
Comment (vmuzikar) [reply docs/guides/server/bootstrap-admin-recovery.adoc:33]: Note that this applies to service account's secret as well.

[the remark this answers is not in the document]
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc:33]: You're correct, I overlooked it for the part in question (as for the service the behaviour should be already documented correctly). I will fix this.

**Comment by vmuzikar** [comment]
@Pepo48 Could you also please address [this comment](https://github.com/keycloak/keycloak/pull/30711#discussion_r1667055511) in this PR since #30711 was merged sooner than this?

File: docs/guides/server/bootstrap-admin-recovery.adoc [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+
+== Bootstrapping an admin service account
+
+In automated scenarios, an admin service account can be a more suitable alternative to a temporary admin user.
+
+To create a temporary admin service account, execute the following command:
+
+<@kc.bootstrapadmin parameters="service"/>
+
+Similarly, the command can be executed regardless of the `master` realm’s existence. Given the client nature of the admin service account, the default admin can be created even after the command execution. If no corresponding environment variables or additional parameters are set, the user will be prompted to enter the required information. For more information, see the <<Environment variables>> section below.
```
[the remark this answers is not in the document]
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc:49]: What I meant by the statement is that you can create the service first and once you open the admin console, you will be still prompted to create the default admin user account. The formulation is weird mainly because at the time of the writing I considered the default admin and the temporary admin to be some sort of different (thus even the NOTE section you rewrote earlier). Only the UI work made me realize that the default admin is also considered to be temporary. Thanks for noticing this, this needs to be rewritten as well.

File: docs/guides/server/bootstrap-admin-recovery.adoc [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+
+== Bootstrapping a temporary admin account at {project_name} startup
+
+{project_name} `start` and `start-dev` commands support options for bootstrapping both temporary admin users and admin service accounts. These options are standard configuration options, so they can be specified in any of the https://www.keycloak.org/server/configuration#_configuring_sources_for_keycloak[configuration sources]. For instance, the following examples demonstrate how to use the `start` and `start-dev` commands with CLI parameters to bootstrap a temporary admin user and an admin service account, respectively:
+
+<@kc.start parameters="--bootstrap-admin-username tmpadm --bootstrap-admin-password pass"/>
+
+<@kc.startdev parameters="--bootstrap-admin-client-id tmpadm --bootstrap-admin-client-secret secret"/>
+
+The purpose of these options is solely for bootstrapping temporary admin accounts. The accounts are always created in the master realm. For recovering lost admin access, use the dedicated command described in the sections below.
```
[the remark this answers is not in the document]
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: Addressed.

File: docs/guides/server/bootstrap-admin-recovery.adoc [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+
+Next, retrieve the `credentialId`. For this example, the OTP credential is the relevant one. Use the following command to get an array of `CredentialRepresentation` objects and find the one with `type` set to `otp`:
+
+<@kc.admin parameters="get users/{userId}/credentials -r {realm}"/>
+
+Finally, the retrieved ID can be used to remove the advanced authentication method (in our case, OTP):
+
+<@kc.admin parameters="delete users/{userId}/credentials/{credentialId} -r {realm}"/>
+
+== Disable the parameters prompt
```
[the remark this answers is not in the document]
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc:83]: Addressed.

File: docs/guides/server/bootstrap-admin-recovery.adoc (hunk 11 above)
[the remark this answers is not in the document]
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: Fixed.

File: docs/guides/server/bootstrap-admin-recovery.adoc (hunk 12 above)
[the remark this answers is not in the document]
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: Addressed by slightly changing the section structure for the dedicated command section of the guide.

File: docs/guides/server/bootstrap-admin-recovery.adoc (hunk 10 above)
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc:99]: I believe this was already partially addressed (I just additionally mentioned the requirement for the server to be stopped) - the question is about the appropriate wording here (see Alex's question and Steven's answer above).

As for the default values, I added a new section.

**Comment by Pepo48** [comment]
> @Pepo48 Could you also please address [this comment](https://github.com/keycloak/keycloak/pull/30711#discussion_r1667055511) in this PR since #30711 was merged sooner than this?

Addressed by adding a new section with the link to the operator's advanced configuration.

File: docs/guides/server/bootstrap-admin-recovery.adoc [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+Additionally, it is strongly recommended to use the dedicated command with the same options that the {project_name} server is started with (e.g., `db` options).
+
+=== Create an admin user
+
+To create a temporary admin user, execute the following command:
+
+<@kc.bootstrapadmin parameters="user"/>
+
+If no other parameters are specified and/or no corresponding environment variables are set, the user is prompted to enter the required information. The username or client ID values can be omitted to use the default values. For more information, see the <<Default values>> and <<Environment variables>> sections below.
```
Comment (vmuzikar) [inline docs/guides/server/bootstrap-admin-recovery.adoc:37]:
```suggestion
If no other parameters are specified and/or no corresponding environment variables are set, the user is prompted to enter the required information. The username value can be omitted to use the default values. For more information, see the <<Default values>> and <<Environment variables>> sections below.
```

File: docs/guides/server/bootstrap-admin-recovery.adoc [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+
+=== Create a service account
+
+In automated scenarios, a temporary admin service account can be a more suitable alternative to a temporary admin user.
+
+To create a temporary admin service account, execute the following command:
+
+<@kc.bootstrapadmin parameters="service"/>
+
+Similarly, if no corresponding environment variables or additional parameters are set, the user will be prompted to enter the required information. The username or client ID values can be omitted to use the default values. For more information, see the <<Default values>> and <<Environment variables>> sections below.
```
Comment (vmuzikar) [inline docs/guides/server/bootstrap-admin-recovery.adoc:53]:
```suggestion
Similarly, if no corresponding environment variables or additional parameters are set, the user will be prompted to enter the required information. The client ID value can be omitted to use the default values. For more information, see the <<Default values>> and <<Environment variables>> sections below.
```

File: docs/guides/server/bootstrap-admin-recovery.adoc [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+
+Alternatively, the parameters can be directly specified in the command:
+
+<@kc.bootstrapadmin parameters="service --client-id tmpclient --client-secret:env=SECRET_VAR"/>
+
+This command creates a temporary admin service account with the client ID `tmpclient` and the secret retrieved from the environment variable.
+
+== Admin bootstrapping using the {project_name} operator
```
Comment (vmuzikar) [inline docs/guides/server/bootstrap-admin-recovery.adoc:61]: I'd do it the other way around to keep it consistent with other guides – instead of add links to server guide pointing to the operator guide, I'd suggest adding a link to the operator guide pointing to server guide.

**Review by vmuzikar — COMMENTED** [review]
@Pepo48 Thanks for another update. Added some last thoughts, mostly nitpics.

File: docs/guides/server/bootstrap-admin-recovery.adoc [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+<#import "/templates/kc.adoc" as kc>
+<#import "/templates/links.adoc" as links>
+
+<@tmpl.guide
+title="Admin bootstrap and recovery"
+summary="Learn how to bootstrap and recover admin account.">
+
+== A temporary admin account
+
+A user or service admin account created using one of the methods described below is *temporary*. This means the account should exist only for the duration necessary to perform operations needed to gain permanent and more secure admin access. Various UI/UX elements, such as the Administration Console warning banner, labels, and log messages, will indicate to a {project_name} administrator that the account is temporary.
```
Comment (mhajas) [inline docs/guides/server/bootstrap-admin-recovery.adoc:11]: Could you elaborate more on what temporary means here? Is the access available until the Keycloak node that was started with the parameters is alive?

File: docs/guides/server/bootstrap-admin-recovery.adoc [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+
+<@kc.startdev parameters="--bootstrap-admin-client-id tmpadm --bootstrap-admin-client-secret secret"/>
+
+The username or client ID values can be omitted; see the <<Default values>> section below for more information.
+
+The purpose of these options is solely for bootstrapping temporary admin accounts. These accounts will be created only during the initial start of the {project_name} server when the master realm doesn't exist yet. The accounts are always created in the master realm. For recovering lost admin access, use the dedicated command described in the sections below.
+
+== Bootstrapping an admin user or service account using the dedicated command
+
+The `bootstrap-admin` command can be executed even before the first-ever start of {project_name}. Bear in mind that the server needs to be stopped prior to using this command. Its execution will trigger the creation of the initial master realm, and as a result, the startup options to bootstrap the admin user and service account will be ignored later when the server is started for the first time.
```
Comment (mhajas) [inline docs/guides/server/bootstrap-admin-recovery.adoc:27]: How does this work with Keycloak cluster? Do all nodes need to be stopped before this procedure is executed?

Comment (vmuzikar) [reply docs/guides/server/bootstrap-admin-recovery.adoc:27]: Yes, all nodes need to be stopped:
https://github.com/keycloak/keycloak/blob/1e14b310669126d2dba3d20c902054f3f253ca61/docs/guides/server/bootstrap-admin-recovery.adoc?plain=1#L27

Maybe we could more emphasis on it, make it a warn box or something.

File: docs/guides/server/bootstrap-admin-recovery.adoc (hunk 18 above)
Comment (vmuzikar) [reply docs/guides/server/bootstrap-admin-recovery.adoc:11]: By "temporary" we mean "considered to be temporary by the user". Basically it's just semantical for now and only highlighted as a [warning](https://github.com/keycloak/keycloak/pull/31387) in the UI and logs. In the future, the accounts will be [expired](https://github.com/keycloak/keycloak/issues/30008#issuecomment-2151956072).

Comment (mhajas) [reply docs/guides/server/bootstrap-admin-recovery.adoc:11]: I see. Should we mention in the docs they need to delete the temporary users manually?

File: docs/guides/server/bootstrap-admin-recovery.adoc (hunk 19 above)
Comment (mhajas) [reply docs/guides/server/bootstrap-admin-recovery.adoc:27]: For example, changing it to "all {project_name} nodes"?

Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: Addressed as per suggestion.

File: docs/guides/server/bootstrap-admin-recovery.adoc (hunk 18 above)
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: Addressed.

File: docs/guides/server/bootstrap-admin-recovery.adoc (hunk 17 above)
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: Addressed.

File: docs/documentation/upgrading/topics/changes/changes-26_0_0.adoc [hunk 20]
```diff
@@ -83,6 +83,8 @@ In order to use a custom footer, create a `footer.ftl` file in your custom login
 
 For more details, see link:{developerguide_link}#_theme_custom_footer[Adding a custom footer to a login theme].
 
-= Admin Bootstrapping
+= Admin Bootstrapping and Recovery
 
-The environment variables `KEYCLOAK_ADMIN` and `KEYCLOAK_ADMIN_PASSWORD` have been deprecated. You should use `KC_BOOTSTRAP_ADMIN_USERNAME` and `KC_BOOTSTRAP_ADMIN_PASSWORD` instead. These are also general options, so they may be specified via the cli or other config sources, for example `--bootstrap-admin-username=admin`.
+It used to be difficult to regain access to a {project_name} instance when all admin users were locked out. The process required multiple advanced steps, including direct database access and manual changes. In an effort to improve the user experience, {project_name} now provides multiple ways to bootstrap a new admin account, which can be used to recover from such situations.
```
[the remark this answers is not in the document]
Comment (Pepo48) [reply docs/documentation/upgrading/topics/changes/changes-26_0_0.adoc:97]: I left the paragraph there and added a new slightly different one to the release notes.

File: docs/guides/server/bootstrap-admin-recovery.adoc (hunk 18 above)
Comment (mhajas) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: Thank you @Pepo48

File: docs/guides/server/bootstrap-admin-recovery.adoc (hunk 19 above)
Comment (mhajas) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: Thank you @Pepo48

**Review by vmuzikar — COMMENTED** [review]
@Pepo48 Thanks for the latest changes. Overall LGTM, added some last minor comments.

@ahus1 @shawkins @mhajas Do you want to re-review as well?

I'd like to merge this rather sooner than later. Not sure we need to wait for a review from @andymunro (as he's quite busy). We could incorporate his review changes as a follow up. WDYT?

File: docs/guides/operator/advanced-configuration.adoc [hunk 21]
```diff
@@ -308,4 +308,12 @@ stringData:
...
 
+=== Admin Bootstrapping
+
+When you create a new instance the Keycloak CR spec.bootstrapAdmin stanza may be used to configure the bootstrap user and/or service account. If you do not specify anything for the spec.bootstrapAdmin, the operator will create a Secret named "metadata.name"-initial-admin with a username temp-admin and a generated password. If you specify a Secret name for bootstrap admin user, then the Secret will need to contain `username` and `password` key value pairs. If you specify a Secret name for bootstrap admin service account, then the Secret will need to contain `client-id` and `client-secret` key value pairs.
+
+If a master realm has already been created for you cluster, then the spec.boostrapAdmin is effectively ignored. If you need to create a recovery admin account, then you'll need to run the CLI command against a Pod directly.
+
+For more information on how to bootstrap a temporary admin user or service account and recover lost admin access, refer to the link:{bootstrapadminrecovery_link}[{bootstrapadminrecovery_name}] guide.
```
[the remark this answers is not in the document]
Comment (Pepo48) [reply docs/guides/operator/advanced-configuration.adoc:317]: what's the general approach here? Should we use a hard-coded links or should we alway try to use variables and link anchors when cross-referencing.

Am I actually allowed to define and use new [document attributes](https://github.com/keycloak/keycloak/pull/31039/files#diff-8fc1f621a47f49335dec8370bcbd8d9e986d76f49442d412c191bc0754a9b75eR64-R65) for this specific case?

[the remark this answers is not in the document]
Comment (vmuzikar) [reply docs/guides/operator/advanced-configuration.adoc:317]: I'm not sure. Possibly @ahus1 has some input?

[the remark this answers is not in the document]
Comment (Pepo48) [reply docs/guides/operator/advanced-configuration.adoc:317]: I guess for the guide itself I can replace it with `<@links.server id="bootstrap-admin-recovery"/>`. But what about the [release notes](https://github.com/keycloak/keycloak/pull/31039/files#diff-0519b2567941e80155425829531006542bea0fbf514b7035e421f3f7a7c936e1R77) then?

**Comment by mhajas** [comment]
@vmuzikar, I went through the changes quickly and it looks good. Still, I am not 100% confident to approve so I will leave it for the others to add final approvals. If you would like me to have a proper look let me know.

File: docs/guides/operator/advanced-configuration.adoc (hunk 21 above)
[the remark this answers is not in the document]
Comment (vmuzikar) [reply docs/guides/operator/advanced-configuration.adoc:317]: Release notes seem to work fine for my local build.

[the remark this answers is not in the document]
Comment (Pepo48) [reply docs/guides/operator/advanced-configuration.adoc:317]: For me as well. The question is whether it's fine to use it that way.

[the remark this answers is not in the document]
Comment (vmuzikar) [reply docs/guides/operator/advanced-configuration.adoc:317]: > For me as well. The question is whether it's fine to use it that way.

For release note and upgrading guide, sure from my PoV. For Operator guide, I'm not sure – locally it doesn't work.

[the remark this answers is not in the document]
Comment (Pepo48) [reply docs/guides/operator/advanced-configuration.adoc:317]: @vmuzikar are the rest of the links in the operator's advanced configuration guide rendered correctly for your local build?

[the remark this answers is not in the document]
Comment (vmuzikar) [reply docs/guides/operator/advanced-configuration.adoc]: Yes, other links render fine for me.

File: docs/guides/operator/advanced-configuration.adoc
[the remark this answers is not in the document]
Comment (vmuzikar) [reply docs/guides/operator/advanced-configuration.adoc:1]: #31646 was now merged. @Pepo48 Can you please rebase?

[the remark this answers is not in the document]
Comment (Pepo48) [reply docs/guides/operator/advanced-configuration.adoc:1]: @vmuzikar rebased.

**Review by vmuzikar — APPROVED** [review]
@Pepo48 After the latest changes, LGTM! Thank you!

@shawkins @ahus1 Do you want to review as well?

**Review by shawkins — APPROVED** [review]
LGTM, thank you @Pepo48
